### ISSUE TITLE
Suggestion: Allow use of const enum keys as keys in interfaces or type aliases

### ISSUE THREAD

**Issue by lostfictions** [Microsoft/TypeScript#16258]
I'm currently writing a small library to make working with the [Web MIDI API](https://www.w3.org/TR/webmidi/) slightly nicer, and I'd like to write event handling code that correctly types the event handler signature (similar to how eg. `lib.dom.d.ts` types [`Document.addEventListener`](https://github.com/Microsoft/TypeScript/blob/451561ec8d793309ca9daa438f22764dd0ef440a/lib/lib.dom.d.ts#L3466) generically with keys/values from the [`DocumentEventMap`](https://github.com/Microsoft/TypeScript/blob/451561ec8d793309ca9daa438f22764dd0ef440a/lib/lib.dom.d.ts#L2675-L2764) interface.)

MIDI events can be dispatched frequently and it's often desirable to handle them with minimum latency, so I'm using const enums to provide an abstraction over MIDI event data (passed in a `UInt8Array` by the Web MIDI API and decoded with bit shifts and masks) without the additional overhead of object lookups.

Since const enums are fully known at design time, it seems like it should be possible to use their keys as keys in interface declarations like `lib.dom.d.ts` does it:

```ts
const enum ChannelMessage {
  'noteoff' = 0x8,
  'noteon' = 0x9,
  'keyaftertouch' = 0xA,
  // ...
}

interface ChannelMessageEventMap {
  [ChannelMessage.noteoff] : NoteEvent
  [ChannelMessage.noteon] : NoteEvent
  [ChannelMessage.keyaftertouch] : AftertouchEvent
  // ...
}

addEventListener<K extends keyof ChannelMessageEventMap>(type: K, listener: (ev: ChannelMessageEventMap[K]) => any): void
```

Unfortunately, this doesn't currently work -- const enum keys can't be used in interfaces. I thought in a pinch I could simply use the enum values as keys in the event map interface instead, that is:

```ts
interface ChannelMessageEventMap {
  0x8 : NoteEvent
  0x9 : NoteEvent
  0xA : AftertouchEvent
  // ...
}
```

The above compiles -- but the following usage of `addEventListener` won't (assuming the signature  declared above):

```ts
addEventListener(ChannelMessage.noteoff, ev => { /* ... */ })
//               ^^^^^^^^^^^^^^^^^^^^^^
// ERROR: Argument of type 'ChannelMessage.'noteoff'' is not assignable to parameter of type '"8" | "9" | "10" | ...
```

Even though const enum values are fully known and inlined at compile time, it seems they aren't compatible with the value literal types themselves.

It's not clear to me if there's any other way to make this work with const enums, unfortunately -- their usefulness as compile-time constants seems greatly reduced by their incompatibility and the inability to use them in other places you'd be able to use a constant value or a type literal. I guess I maybe have to bite the bullet and map to string keys after all?

**Comment by RyanCavanaugh** [comment]
Similar to #5579 but it might be more tractable to do this only for keys which originate in enums

**Comment by jsen-** [comment]
I'm running into similar issue, trying to create function translating HTTP status codes to messages.
Something like ([playground](https://www.typescriptlang.org/play/#src=const%20STATUS_CODES%20%3D%20%7B%0D%0A%20%20%20%20404%3A%20%22Not%20found%22%2C%0D%0A%20%20%20%20500%3A%20%22Internal%20server%20error%22%2C%0D%0A%7D%3B%0D%0A%0D%0Atype%20NumericKeys%3CT%20extends%20%7B%20%5Bkey%3A%20number%5D%3A%20string%20%7D%3E%20%3D%20%7B%0D%0A%20%20%20%20%5BP%20in%20keyof%20T%5D%3A%20string%3B%0D%0A%7D%3B%0D%0A%0D%0Afunction%20code2msg(code%3A%20keyof%20NumericKeys%3Ctypeof%20STATUS_CODES%3E)%20%7B%0D%0A%20%20%20%20return%20STATUS_CODES%5Bcode%5D%3B%0D%0A%7D%0D%0A%0D%0Acode2msg(%22404%22)%3B%0D%0Acode2msg(404)%3B%0D%0A))
```typescript
const STATUS_CODES = {
    404: "Not found",
    500: "Internal server error",
};

type NumericKeys<T extends { [key: number]: string }> = {
    [P in keyof T]: string;
};

function code2msg(code: keyof NumericKeys<typeof STATUS_CODES>) {
    return STATUS_CODES[code];
}

code2msg("404"); // works
code2msg(404); // fails
```
It fails due to `keyof NumericKeys<typeof STATUS_CODES>` is of type "404" | "500" instead of `404 | 500`.
Which in turn is due to `keyof` being always of type `string` even when the constraint of ` NumericKeys<T extends { [key: number]: string }>` is used.

Would it be possible for `keyof` to be constrained by it's outer generic type (instead of always string)?
Motivation is to encode in the type system that `code2msg` can only be called on supported status codes.

**Comment by RyanCavanaugh** [comment]
@jsen- I think the behavior there is really what you want (or at least *should* want, if that's a thing). Consider if you had written something like this:
```ts
function code2msg(code: keyof NumericKeys<typeof STATUS_CODES>) {
    for (var k of STATUS_CODES) {
      if (k === code) return STATUS_CODES[k];
    }
    return undefined;
}
```
which looks like it should do the same thing, but doesn't. Since we don't really know what you're going to use a `keyof` *for*, the only safe thing is to always say `string`.

**Comment by jsen-** [comment]
I believe you meant
```typescript
for (var k in STATUS_CODES) {
```
Under my logic, `k === code` should be a compile time error comparing `k: string` with `code: 404 | 500`

**Comment by Kovensky** [comment]
This affects mapped types in more natural use cases as well.

```ts
const enum Foo {
  Bar = 'bar'
}

type FooMap = {[key in Foo]: boolean}

const foo: FooMap = {
  [Foo.Bar]: true
}
```
```
Type '{ [x: string]: boolean; }' is not assignable to type 'FooMap'.
Property 'bar' is missing in type '{ [x: string]: boolean; }'.
```

I also seem to not be able to index the mapped type _but_ I can't reproduce it in a small example; I assume it only happens if the `enum` is imported from somewhere else.

```ts
const isBar = foo[Foo.Bar] // not an error in this example, but errors out with missing index signature in a more complex case
```

**Comment by michaeljota** [comment]
Actually it works as expect when compiled, it's just the type checking that broke and get lost. Tested in Playground with `number` enums only, but it should also work with `string` enums if the compiled object it's the same one.

**Comment by mhegazy** [comment]
Should be covered by #5579

**Comment by begincalendar** [comment]
Is it possible to not conflate this issue with #5579?

I get that they are related, but given that #5579 has been open for nearly two years and suffers from complexity such as mentioned by Ryan [here](https://stackoverflow.com/a/41403082/2117581), is it possible to assess the feasibility of making this happen for constant expressions (at least for now)?

**Comment by sanex3339** [comment]
Really want this feature because i don't like that:
https://github.com/javascript-obfuscator/javascript-obfuscator/blob/5d5a9a9ee1d8fc312b4fc85c4baedc1a3f0d7db2/src/utils/Utils.ts#L70

**Comment by mhegazy** [comment]
> I get that they are related, but given that #5579 has been open for nearly two years and suffers from complexity such as mentioned by Ryan here, is it possible to assess the feasibility of making this happen for constant expressions (at least for now)?

we have a fix for #5579 in https://github.com/Microsoft/TypeScript/pull/15473. so do not think we need another issue for it.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
